File: Simulation/simulation.py
import networkx as nx
from random import *
from collections import *

def rand_fixed_degree_graph(l, r, d):
  # create nodes
  left_nodes  = ["l"+str(i).zfill(4) for i in range(l)]
  right_nodes = ["r"+str(i).zfill(5) for i in range(r)]
  g = nx.Graph()
  g.add_nodes_from(left_nodes , bipartite="left")
  g.add_nodes_from(right_nodes, bipartite="right")

  # create edges
  for u in left_nodes:
    for sample_index in sample(xrange(r), d):
      g.add_edge(u, right_nodes[sample_index])

  return g

def sampling_algorithm(g, c, a):
  hit_count = defaultdict(int)
  left_vertices = (u for u in g if g.node[u]["bipartite"]=="left")

  for u in left_vertices:
    neighbors = g.neighbors(u)
    if len(neighbors) < c:
      samples = neighbors
    else:
      samples = [neighbors[i] for i in sample(xrange(len(neighbors)), c)]

    for sample_ in samples:
      hit_count[sample_] += 1

  return [v for v in hit_count if hit_count[v] >= a]

def greedy_algorithm(g, c, a):
  degree_count = defaultdict(int)
  right_vertices = (u for u in g if g.node[u]["bipartite"]=="right")
  solution = []

  for v in right_vertices:
    eligible_neighbors = [u for u in g.neighbors(v) if degree_count[u] < c]
    if len(eligible_neighbors) >= a:
      for u in eligible_neighbors[:a]:
        degree_count[u] += 1
      solution.append(v)

  return solution

def karp_sipser(g, c, a):
  gg = g.copy()
  left_vertices  = (u for u in gg if gg.node[u]["bipartite"]=="left")
  right_vertices = (v for v in gg if gg.node[v]["bipartite"]=="right")
  
  left_used_degree = defaultdict(int)
  right_rem_degree = defaultdict(set)
  for v in right_vertices:
    if gg.degree(v) >= a:
      right_rem_degree[gg.degree(v)].add(v)

  solution = []
  try:
    min_degree = min(d for d in right_rem_degree if len(right_rem_degree[d])>0)
  except:
    min_degree = 0

  while min_degree >= a:
    # pick vertex on the right with lowest degree and remove it
    v = sample(right_rem_degree[min_degree],1)[0]
    right_rem_degree[min_degree].remove(v)
    eligible_neighbors = gg.neighbors(v)
    gg.remove_node(v)

    # pick a neighbors. if they become saturated, remove them
    for u in eligible_neighbors[:a]:
      left_used_degree[u] += 1
      if left_used_degree[u] == c:
        second_neighbors = gg.neighbors(u)
        for vv in second_neighbors:
          prev_degree = gg.degree(vv)
          if prev_degree < a: continue

          right_rem_degree[prev_degree].remove(vv)
          if prev_degree > a:
            right_rem_degree[prev_degree-1].add(vv)
        gg.remove_node(u)

    solution.append(v)
    degrees = [d for d in right_rem_degree if len(right_rem_degree[d])>0]
    if len(degrees)==0: break
    min_degree = min(degrees)

  return solution  

# parameters
trials = 10
l      = 10000
r      = 100000
d      = 10

for a in range(1,10):
  for c in range(1,10):
    optimal = min(r, c*l/a)
    for _ in xrange(trials):
      g = rand_fixed_degree_graph(l, r, d)
      s1 = sampling_algorithm(g, c, a)
      s2 = greedy_algorithm(g, c, a)
      s3 = karp_sipser(g, c, a)
      print "c = %d, a = %d" % (c, a)
      print "Sampling    : %d" % (len(s1))
      print "Greedy      : %d" % (len(s2))
      print "Karp-Sipser : %d" % (len(s3))
      print "Optimal     : %d" % (optimal)
      print 

File: Simulation/simulation.scala
import scala.io.Source
import scala.util.Random
import scala.collection.mutable.Map
import scala.collection.mutable.{ArrayBuffer => AB}

class Graph[T](
  V: Seq[T],
  E: Map[T, _ <: Seq[T]]) {

  def this() = this(List(), Map())
  def this(V: Seq[T]) = this(V, Map()) 

  val vertices: AB[T] = AB(V: _*)
  val edges   : Map[T, AB[T]] = Map()
  V foreach { v => 
    edges += v -> {if (E.contains(v)) AB(E(v): _*) else AB() }
  }
  
  def degree(v: T) = {
    edges(v).length
  }

  def addVertex(v: T, neighbors: Seq[T] = AB()) = {
    edges += (v->AB())
    vertices += v
    neighbors foreach {u => 
      edges(v) += u
      edges(u) += v
    }
  }
  
  def removeVertex(v: T) = {
    val neighbors = edges(v)
    neighbors foreach { edges(_) -= v }
    edges -= v
  }

  def addEdge(v: T, u: T) = {
    edges(v) += u
    edges(u) += v
  }
  
  def removeEdge(v: T, u: T) = {
    edges(v) -= u
    edges(u) -= v
  }
  
  override def clone() = {
    new Graph(vertices, edges)
  }
}

def samplingAlgorithm[T <: Int](g: Graph[T], c: Int, a: Int): Int = {
  val hitCount: Map[T, Int] = Map().withDefaultValue(0)
  val leftVertices = g.vertices filter {_ > 0}
  leftVertices foreach { u =>
    val neighbors = g.edges(u)
    sample(neighbors, c) foreach {hitCount(_) += 1}
  }
  var total = 0
  hitCount.keySet.foreach {v => if (hitCount(v) >= a) total += 1}
  return total
}

def greedyAlgorithm[T <: Int](g: Graph[T], c: Int, a: Int): Int = {
  val usedDegree: Map[T, Int] = Map().withDefaultValue(0)
  val rightVertices = g.vertices filter { _ < 0}
  val solution: AB[Int] = AB()

  rightVertices foreach { v => 
    val eligibleNeighbors = g.edges(v).filter(usedDegree(_) < c)
    if (eligibleNeighbors.length >= a) {
      (0 until a) foreach {t => usedDegree(eligibleNeighbors(t)) += 1}
      solution += v
    }
  }
  return solution.length
}

def iteratingAlgorithm[T <: Int](gg: Graph[T], c: Int, a: Int): Int = {
  val g = gg.clone
  val usedDegree: Map[T, Int] = Map().withDefaultValue(0)
  val leftVertices  = g.vertices filter { _ > 0}
  val rightVertices = g.vertices filter { _ < 0}
  val buckets       = AB(AB(rightVertices:_*))

  
  (1 to a) foreach { i =>
    buckets += AB()
    buckets(i-1) foreach { v =>
      val eligibleNeighbors = g.edges(v) filter {usedDegree(_) < c}
      if (!eligibleNeighbors.isEmpty) {
        val u = eligibleNeighbors minBy (usedDegree(_))
        g.removeEdge(v,u)
        buckets(i) += v
        usedDegree(u) += 1
      }
    }
  }
  return buckets(a).length
}

val r = new Random

def genRandomRegularGraphInt(l: Int, r: Int, d: Int): Graph[Int] = {
  val leftVertices  = 1 to l
  val rightVertices = 1 to r map {-_}
  val G = new Graph(rightVertices)

  leftVertices foreach { u => G.addVertex(u, sample(rightVertices, d)) }
  return G
}

def sample[T](l: IndexedSeq[T], count: Int = 1): Seq[T] = {
  val choices: AB[Int] = AB()
  while (choices.length < count) {
    val nextChoice = r.nextInt(l.length)
    if (!choices.contains(nextChoice) )
      choices += nextChoice}
  choices map {l(_)}
}


def benchmark(
  c: Int, a: Int,
  algo: (Graph[Int], Int, Int)=>Int,
  trials: Int = 10): Int = {
  
  val results = (1 to trials) map { t =>
    val G = genRandomRegularGraphInt(1000,10000,50)
    algo(G, c, a)
  }
  return results.sum / results.length
}

for (c <- 1 until 10) {
  for (a <- 1 until 10) {
    println("c=%d, a=%d".format(c, a))
    println("Greedy   : %d".format(benchmark(c, a, greedyAlgorithm)))
    println("Sampling : %d".format(benchmark(c, a, samplingAlgorithm)))
    println("CTO      : %d".format(benchmark(c, a, iteratingAlgorithm)))
    println
  }
}